From cdbc0176a450f26df75a4d1c900c9ff5fd206311 Mon Sep 17 00:00:00 2001
From: Michael Stack <stack@apache.org>
Date: Mon, 4 Dec 2017 20:58:02 -0800
Subject: [PATCH] HBASE-19295 The Configuration returned by CPEnv should be
 read-only.

Adds a ReadOnlyConfiguration that delegates gets but throws exception
on sets/adds, etc.
---
 .../hadoop/hbase/CoprocessorEnvironment.java       |   2 +-
 .../hadoop/hbase/coprocessor/BaseEnvironment.java  |   2 +-
 .../hbase/coprocessor/ReadOnlyConfiguration.java   | 366 +++++++++++++++++++++
 .../coprocessor/TestCoprocessorConfiguration.java  |  34 ++
 4 files changed, 402 insertions(+), 2 deletions(-)
 create mode 100644 hbase-server/src/main/java/org/apache/hadoop/hbase/coprocessor/ReadOnlyConfiguration.java

diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/CoprocessorEnvironment.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/CoprocessorEnvironment.java
index 418d6247eb..8678e1d0a7 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/CoprocessorEnvironment.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/CoprocessorEnvironment.java
@@ -45,7 +45,7 @@ public interface CoprocessorEnvironment<C extends Coprocessor> {
   /** @return the load sequence number */
   int getLoadSequence();
 
-  /** @return the configuration */
+  /** @return a Read-only Configuration */
   Configuration getConfiguration();
 
   /**
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/coprocessor/BaseEnvironment.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/coprocessor/BaseEnvironment.java
index ebbca65784..9f5ca231f5 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/coprocessor/BaseEnvironment.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/coprocessor/BaseEnvironment.java
@@ -57,7 +57,7 @@ public class BaseEnvironment<C extends Coprocessor> implements CoprocessorEnviro
     this.priority = priority;
     this.state = Coprocessor.State.INSTALLED;
     this.seq = seq;
-    this.conf = conf;
+    this.conf = new ReadOnlyConfiguration(conf);
   }
 
   /** Initialize the environment */
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/coprocessor/ReadOnlyConfiguration.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/coprocessor/ReadOnlyConfiguration.java
new file mode 100644
index 0000000000..0cded8ee67
--- /dev/null
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/coprocessor/ReadOnlyConfiguration.java
@@ -0,0 +1,366 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.coprocessor;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.HBaseInterfaceAudience;
+import org.apache.hadoop.io.Writable;
+import org.apache.yetus.audience.InterfaceAudience;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Reader;
+import java.io.Writer;
+import java.net.InetSocketAddress;
+import java.net.URL;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+/**
+ * Wraps a Configuration to make it read-only.
+ */
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.COPROC)
+class ReadOnlyConfiguration extends Configuration {
+  private final Configuration conf;
+
+  ReadOnlyConfiguration(final Configuration conf) {
+    this.conf = conf;
+  }
+
+  public void setDeprecatedProperties() {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void addResource(String name) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void addResource(URL url) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void addResource(Path file) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void addResource(InputStream in) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void addResource(InputStream in, String name) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void addResource(Configuration conf) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void reloadConfiguration() {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public String get(String name) {
+    return conf.get(name);
+  }
+
+  public void setAllowNullValueProperties(boolean val) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public boolean onlyKeyExists(String name) {
+    return conf.onlyKeyExists(name);
+  }
+
+  public String getTrimmed(String name) {
+    return conf.getTrimmed(name);
+  }
+
+  public String getTrimmed(String name, String defaultValue) {
+    return conf.getTrimmed(name, defaultValue);
+  }
+
+  public String getRaw(String name) {
+    return conf.getRaw(name);
+  }
+
+  public void set(String name, String value) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void set(String name, String value, String source) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void unset(String name) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void setIfUnset(String name, String value) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public String get(String name, String defaultValue) {
+    return conf.get(name, defaultValue);
+  }
+
+  public int getInt(String name, int defaultValue) {
+    return conf.getInt(name, defaultValue);
+  }
+
+  public int[] getInts(String name) {
+    return conf.getInts(name);
+  }
+
+  public void setInt(String name, int value) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public long getLong(String name, long defaultValue) {
+    return conf.getLong(name, defaultValue);
+  }
+
+  public long getLongBytes(String name, long defaultValue) {
+    return conf.getLongBytes(name, defaultValue);
+  }
+
+  public void setLong(String name, long value) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public float getFloat(String name, float defaultValue) {
+    return conf.getFloat(name, defaultValue);
+  }
+
+  public void setFloat(String name, float value) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public double getDouble(String name, double defaultValue) {
+    return conf.getDouble(name, defaultValue);
+  }
+
+  public void setDouble(String name, double value) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public boolean getBoolean(String name, boolean defaultValue) {
+    return conf.getBoolean(name, defaultValue);
+  }
+
+  public void setBoolean(String name, boolean value) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void setBooleanIfUnset(String name, boolean value) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public <T extends Enum<T>> void setEnum(String name, T value) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public <T extends Enum<T>> T getEnum(String name, T defaultValue) {
+    return conf.getEnum(name, defaultValue);
+  }
+
+  public void setTimeDuration(String name, long value, TimeUnit unit) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public long getTimeDuration(String name, long defaultValue, TimeUnit unit) {
+    return conf.getTimeDuration(name, defaultValue, unit);
+  }
+
+  public Pattern getPattern(String name, Pattern defaultValue) {
+    return conf.getPattern(name, defaultValue);
+  }
+
+  public void setPattern(String name, Pattern pattern) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public String[] getPropertySources(String name) {
+    return conf.getPropertySources(name);
+  }
+
+  public Configuration.IntegerRanges getRange(String name, String defaultValue) {
+    return conf.getRange(name, defaultValue);
+  }
+
+  public Collection<String> getStringCollection(String name) {
+    return conf.getStringCollection(name);
+  }
+
+  public String[] getStrings(String name) {
+    return conf.getStrings(name);
+  }
+
+  public String[] getStrings(String name, String... defaultValue) {
+    return conf.getStrings(name, defaultValue);
+  }
+
+  public Collection<String> getTrimmedStringCollection(String name) {
+    return conf.getTrimmedStringCollection(name);
+  }
+
+  public String[] getTrimmedStrings(String name) {
+    return conf.getTrimmedStrings(name);
+  }
+
+  public String[] getTrimmedStrings(String name, String... defaultValue) {
+    return conf.getTrimmedStrings(name, defaultValue);
+  }
+
+  public void setStrings(String name, String... values) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public char[] getPassword(String name) throws IOException {
+    return conf.getPassword(name);
+  }
+
+  public InetSocketAddress getSocketAddr(String hostProperty, String addressProperty, String defaultAddressValue, int defaultPort) {
+    return conf.getSocketAddr(hostProperty, addressProperty, defaultAddressValue, defaultPort);
+  }
+
+  public InetSocketAddress getSocketAddr(String name, String defaultAddress, int defaultPort) {
+    return conf.getSocketAddr(name, defaultAddress, defaultPort);
+  }
+
+  public void setSocketAddr(String name, InetSocketAddress addr) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public InetSocketAddress updateConnectAddr(String hostProperty, String addressProperty, String defaultAddressValue, InetSocketAddress addr) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public InetSocketAddress updateConnectAddr(String name, InetSocketAddress addr) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public Class<?> getClassByName(String name) throws ClassNotFoundException {
+    return conf.getClassByName(name);
+  }
+
+  public Class<?> getClassByNameOrNull(String name) {
+    return conf.getClassByNameOrNull(name);
+  }
+
+  public Class<?>[] getClasses(String name, Class<?>[] defaultValue) {
+    return conf.getClasses(name, defaultValue);
+  }
+
+  public Class<?> getClass(String name, Class<?> defaultValue) {
+    return conf.getClass(name, defaultValue);
+  }
+
+  public <U> Class<? extends U> getClass(String name, Class<? extends U> defaultValue, Class<U> xface) {
+    return conf.getClass(name, defaultValue, xface);
+  }
+
+  public <U> List<U> getInstances(String name, Class<U> xface) {
+    return conf.getInstances(name, xface);
+  }
+
+  public void setClass(String name, Class<?> theClass, Class<?> xface) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public Path getLocalPath(String dirsProp, String path) throws IOException {
+    return conf.getLocalPath(dirsProp, path);
+  }
+
+  public File getFile(String dirsProp, String path) throws IOException {
+    return conf.getFile(dirsProp, path);
+  }
+
+  public URL getResource(String name) {
+    return conf.getResource(name);
+  }
+
+  public InputStream getConfResourceAsInputStream(String name) {
+    return conf.getConfResourceAsInputStream(name);
+  }
+
+  public Reader getConfResourceAsReader(String name) {
+    return conf.getConfResourceAsReader(name);
+  }
+
+  public Set<String> getFinalParameters() {
+    return conf.getFinalParameters();
+  }
+
+  public int size() {
+    return conf.size();
+  }
+
+  public void clear() {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public Iterator<Map.Entry<String, String>> iterator() {
+    return conf.iterator();
+  }
+
+  public void writeXml(OutputStream out) throws IOException {
+    conf.writeXml(out);
+  }
+
+  public void writeXml(Writer out) throws IOException {
+    conf.writeXml(out);
+  }
+
+  public ClassLoader getClassLoader() {
+    return conf.getClassLoader();
+  }
+
+  public void setClassLoader(ClassLoader classLoader) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public String toString() {
+    return conf.toString();
+  }
+
+  public void setQuietMode(boolean quietmode) {
+    throw new IllegalStateException("Read-only Configutation");
+  }
+
+  public void readFields(DataInput in) throws IOException {
+    conf.readFields(in);
+  }
+
+  public void write(DataOutput out) throws IOException {
+    conf.write(out);
+  }
+
+  public Map<String, String> getValByRegex(String regex) {
+    return conf.getValByRegex(regex);
+  }
+}
diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestCoprocessorConfiguration.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestCoprocessorConfiguration.java
index 6213e86bf5..a1e5e0fd58 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestCoprocessorConfiguration.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestCoprocessorConfiguration.java
@@ -20,6 +20,7 @@
 package org.apache.hadoop.hbase.coprocessor;
 
 import java.io.IOException;
+import java.util.List;
 import java.util.Optional;
 import java.util.concurrent.atomic.AtomicBoolean;
 
@@ -27,6 +28,7 @@ import static org.mockito.Mockito.*;
 import static org.junit.Assert.*;
 
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.Coprocessor;
 import org.apache.hadoop.hbase.CoprocessorEnvironment;
 import org.apache.hadoop.hbase.HBaseConfiguration;
 import org.apache.hadoop.hbase.HRegionInfo;
@@ -170,4 +172,36 @@ public class TestCoprocessorConfiguration {
     assertFalse("Table coprocessors should not have been loaded",
       tableCoprocessorLoaded.get());
   }
+
+  /**
+   * Rough test that Coprocessor Environment is Read-Only.
+   * Just check a random CP and see that it returns a read-only config.
+   * @throws Exception
+   */
+  @Test
+  public void testReadOnlyConfiguration() throws Exception {
+    Configuration conf = new Configuration(CONF);
+    HRegion region = mock(HRegion.class);
+    when(region.getRegionInfo()).thenReturn(REGIONINFO);
+    when(region.getTableDescriptor()).thenReturn(TABLEDESC);
+    RegionServerServices rsServices = mock(RegionServerServices.class);
+    RegionCoprocessorHost rcp = new RegionCoprocessorHost(region, rsServices, conf);
+    boolean found = false;
+    for (String cpStr: rcp.getCoprocessors()) {
+      CoprocessorEnvironment cpenv = rcp.findCoprocessorEnvironment(cpStr);
+      if (cpenv != null) {
+        found = true;
+      }
+      Configuration c = cpenv.getConfiguration();
+      boolean expected = false;
+      try {
+        c.set("one.two.three", "four.five.six");
+        fail("Should be read-only Configuration");
+      } catch (IllegalStateException ise) {
+        expected = true;
+      }
+      assertTrue("Failed to get expected exception", expected);
+    }
+    assertTrue("Should be at least one CP found", found);
+  }
 }
-- 
2.11.0 (Apple Git-81)
